File: packages/sdk/src/types/export-import.ts
/**
 * Export/Import Types (SDK)
 *
 * Mirrored from packages/engram/src/types/export-import.ts.
 * Keep in sync with the backend source of truth.
 * Date fields use `string` (JSON wire format, not Date objects).
 *
 * Naming difference: SDK uses `ImportOptions` where the backend uses
 * `ImportParams` — a deliberate rename for public API clarity.
 */

// ============================================================================
// Export Types
// ============================================================================

/**
 * Which data domains to include in an export.
 * Note: "knowledge" and "crystals" scopes both map to the unified
 * knowledge_crystals table on the server (ADR-055).
 */
export type ExportScope = "knowledge" | "crystals" | "sessions";

/**
 * Entity types used in export/import operations.
 * Unified types replace the former separate knowledge/crystal entity names.
 */
export type ExportEntityType =
  | "knowledge_crystals"
  | "knowledge_crystal_edges"
  | "crystal_versions"
  | "crystal_acl"
  | "crystal_share_links"
  | "sessions"
  | "session_notes"
  | "session_note_edges"
  | "session_constraints"
  | "session_links"
  | "decision_points"
  | "exploration_branches"
  | "stuck_detections"
  // Deprecated aliases (ADR-057)
  | "knowledge_items"
  | "knowledge_edges"
  | "crystals"
  | "crystal_memberships"
  | "crystal_edges";

/** Filters to narrow the set of entities included in an export. */
export interface ExportFilter {
  // Time filters
  since?: string;
  until?: string;

  // Source filters
  crystalIds?: string[];
  sessionIds?: string[];
  sourceProject?: string;

  // Type filters
  /** @deprecated Use nodeTypes instead */
  types?: string[];
  /** @deprecated Use nodeTypes instead */
  crystalTypes?: string[];
  /** Unified node type filter (ADR-055) */
  nodeTypes?: string[];

  // Quality filters
  verified?: boolean;
  minConfidence?: number;
  embedded?: boolean;

  // Relationship filters
  includeRelatedDepth?: number;
}

/** Parameters for initiating an export or estimating its size. */
export interface ExportParams {
  scopes: ExportScope[];
  filters: ExportFilter;
  format: "ndjson" | "archive";
  compress: boolean;
}

/** Estimated entity and byte counts for a prospective export. */
export interface ExportEstimate {
  /** @deprecated Use knowledgeCrystals instead */
  knowledgeItems?: number;
  /** @deprecated Use knowledgeCrystalEdges instead */
  knowledgeEdges?: number;
  /** @deprecated Use knowledgeCrystals instead */
  crystals?: number;
  /** @deprecated Use knowledgeCrystalEdges instead */
  crystalMemberships?: number;
  /** Unified node count (replaces knowledgeItems + crystals) */
  knowledgeCrystals: number;
  /** Unified edge count (replaces knowledgeEdges + crystalMemberships) */
  knowledgeCrystalEdges: number;
  sessions: number;
  sessionNotes: number;
  totalEntities: number;
  estimatedSizeBytes: number;
}

// ============================================================================
// Import Types
// ============================================================================

/** Strategy for resolving conflicts during import. */
export type ConflictResolution = "newer" | "skip" | "overwrite" | "prompt";

/** Options controlling import behaviour. */
export interface ImportOptions {
  onConflict: ConflictResolution;
  wipe?: boolean;
  wipeAll?: boolean;
  preview?: boolean;
  force?: boolean;
  ignoreChecksums?: boolean;
  targetProject?: string;
}

/** A single entity conflict detected during import preview. */
export interface ImportConflict {
  id: string;
  entityType: string;
  title: string;
  localUpdatedAt: string;
  importUpdatedAt: string;
}

/** Result of a preview import run — reports what would happen without changes. */
export interface ImportPreview {
  success: boolean;
  schemaVersion?: {
    archive: string;
    current: string;
    migrationRequired: boolean;
  };
  counts?: Record<string, { new: number; updated: number; skipped: number }>;
  conflicts?: ImportConflict[];
  conflictCount?: number;
  error?: {
    code: string;
    message: string;
  };
}

/** Result of a completed import operation. */
export interface ImportResult {
  success: boolean;
  counts: Record<string, { inserted: number; updated: number; skipped: number }>;
  errors: Array<{ entityType: string; id: string; error: string }>;
  duration: number;
}

File: packages/sdk-python/engram/types/export_import.py
"""Export/Import types for the Engram SDK."""
from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

__all__ = [
    # Enums / Literal types
    "ExportScope",
    "ExportFormat",
    "ConflictResolution",
    # Request param models
    "ExportFilter",
    "ExportParams",
    "ImportOptions",
    # Response models
    "ExportEstimate",
    "ImportConflict",
    "ImportPreview",
    "ImportPreviewSchemaVersion",
    "ImportPreviewCounts",
    "ImportPreviewError",
    "ImportResult",
    "ImportResultError",
    "ImportResultCounts",
]

# ============================================================================
# Enums / Literal Types
# ============================================================================

ExportScope = Literal["knowledge", "crystals", "sessions"]

ExportFormat = Literal["ndjson", "archive"]

ConflictResolution = Literal["newer", "skip", "overwrite", "prompt"]

# ============================================================================
# Request Param Models (camelCase — server Zod schemas use camelCase)
# ============================================================================


class ExportFilter(BaseModel):
    """Filters for export requests."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    since: Optional[str] = None
    until: Optional[str] = None
    crystal_ids: Optional[list[str]] = None
    session_ids: Optional[list[str]] = None
    source_project: Optional[str] = None
    types: Optional[list[str]] = None
    crystal_types: Optional[list[str]] = None
    verified: Optional[bool] = None
    min_confidence: Optional[float] = None
    embedded: Optional[bool] = None
    include_related_depth: Optional[int] = None


class ExportParams(BaseModel):
    """Request params for starting an export or estimating export size."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    scopes: list[ExportScope]
    filters: Optional[ExportFilter] = None
    format: ExportFormat = "ndjson"
    compress: bool = True


class ImportOptions(BaseModel):
    """Options for import operations."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    on_conflict: ConflictResolution = "newer"
    wipe: Optional[bool] = None
    wipe_all: Optional[bool] = None
    preview: Optional[bool] = None
    force: Optional[bool] = None
    ignore_checksums: Optional[bool] = None
    target_project: Optional[str] = None


# ============================================================================
# Response Models
# ============================================================================


class ExportEstimate(BaseModel):
    """Estimate of export size and entity counts."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    knowledge_items: int
    knowledge_edges: int
    crystals: int
    crystal_memberships: int
    sessions: int
    session_notes: int
    total_entities: int
    estimated_size_bytes: int


class ImportConflict(BaseModel):
    """A conflict detected during import preview."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    id: str
    entity_type: str
    title: str
    local_updated_at: str
    import_updated_at: str


class ImportPreviewSchemaVersion(BaseModel):
    """Schema version information from import preview."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    archive: str
    current: str
    migration_required: bool


class ImportPreviewCounts(BaseModel):
    """Counts for a single entity type in import preview."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    new: int
    updated: int
    skipped: int


class ImportPreviewError(BaseModel):
    """Error information from import preview."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    code: str
    message: str


class ImportPreview(BaseModel):
    """Result of previewing an import."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    success: bool
    schema_version: Optional[ImportPreviewSchemaVersion] = None
    counts: Optional[dict[str, ImportPreviewCounts]] = None
    conflicts: Optional[list[ImportConflict]] = None
    conflict_count: Optional[int] = None
    error: Optional[ImportPreviewError] = None


class ImportResultCounts(BaseModel):
    """Counts for a single entity type in import result."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    inserted: int
    updated: int
    skipped: int


class ImportResultError(BaseModel):
    """An error for a single entity during import."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    entity_type: str
    id: str
    error: str


class ImportResult(BaseModel):
    """Result of an import operation."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    success: bool
    counts: dict[str, ImportResultCounts]
    errors: list[ImportResultError]
    duration: float
